### ISSUE TITLE
Botocore endpoint timeout not the same as the lambda timeout

### ISSUE THREAD

**Issue by lucasdf** [aws/aws-cli#2657]
**Issue:**

I have an AWS Lambda function that is expected to run for a few minutes and it has a timeout of 4 minutes. I noticed that when I invoke this function using the CLI the function is always invoked multiple times, basically once every one minute. 

By using the "--debug" flag I noticed that botocore is setting the timeout of the endpoint as 60 seconds. So this is the cause of the multiple invocations: since the CLI reaches the timeout after one minute it sends a new request every minute, despite the fact that the lambda function's own timeout is higher than this value and the first invocation is still being processed.

**Questions:**
1. What is the expected behaviour of botocore/aws-cli: is it to use 60 seconds as timeout no matter what the lambda function timeout is, or is the expected behaviour to get the timeout value from the function definition? It seems to me that using the same timeout as the function should be the expected behaviour. If indeed it is then there seems to be something wrong when setting this value.
2. Is there a way to set the timeout value while invoking `aws lambda invoke`? `aws lambda invoke help` does not display any information regarding this.

**Commands:**
I am invoking the function using the following command:
`aws lambda invoke --function-name FUNCTION_NAME --payload 'JSON_PAYLOAD' output.json`
The debug flag shows the following message:
`MainThread - botocore.endpoint - DEBUG - Setting lambda timeout as (60, 60)`

**Versions:**
`aws-cli/1.11.93 Python/2.7.12 Linux/4.4.0-79-generic botocore/1.5.56`
`My lambda function is using Python 3.6`

**Reproducing the issue:**
It should be possible to reproduce it by creating a lambda function that just sleeps for more than 60 seconds and invoking it by using the CLI. 

**tl;dr:**
My function is being invoked multiple times because botocore is using 60 seconds as the timeout instead of using the four minutes that is defined in my lambda function. The expected behaviour should be to use the same timeout used by the lambda function. At least there should be a way for me to define the timeout value as a parameter when using `aws lambda invoke`.

**Comment by lucasdf** [comment]
Full log: the function is invoked at 22:32:13. At 22:33:14 the timeout occur and the function is invoked again.

lucas@lucas-ubuntu:~/projects/app-shogun/app_shogun_lambda$ aws lambda invoke --debug --function-name test --payload '{"test_sleep": "true", "test_sleep_val": 80 }' output.json
2017-06-10 22:32:13,415 - MainThread - awscli.clidriver - DEBUG - CLI version: aws-cli/1.11.93 Python/2.7.12 Linux/4.4.0-79-generic botocore/1.5.56
2017-06-10 22:32:13,415 - MainThread - awscli.clidriver - DEBUG - Arguments entered to CLI: ['lambda', 'invoke', '--debug', '--function-name', 'test', '--payload', '{"test_sleep": "true", "test_sleep_val": 80 }', 'output.json']
2017-06-10 22:32:13,415 - MainThread - botocore.hooks - DEBUG - Event session-initialized: calling handler <function add_scalar_parsers at 0x7f4c417bb5f0>
2017-06-10 22:32:13,415 - MainThread - botocore.hooks - DEBUG - Event session-initialized: calling handler <function inject_assume_role_provider_cache at 0x7f4c423289b0>
2017-06-10 22:32:13,417 - MainThread - botocore.loaders - DEBUG - Loading JSON file: /usr/local/lib/python2.7/dist-packages/botocore/data/lambda/2015-03-31/service-2.json
2017-06-10 22:32:13,431 - MainThread - botocore.hooks - DEBUG - Event service-data-loaded.lambda: calling handler <function register_retries_for_service at 0x7f4c427692a8>
2017-06-10 22:32:13,431 - MainThread - botocore.handlers - DEBUG - Registering retry handlers for service: lambda
2017-06-10 22:32:13,432 - MainThread - botocore.hooks - DEBUG - Event building-command-table.lambda: calling handler <function add_waiters at 0x7f4c417c29b0>
2017-06-10 22:32:13,436 - MainThread - awscli.clidriver - DEBUG - OrderedDict([(u'function-name', <awscli.arguments.CLIArgument object at 0x7f4c413bfbd0>), (u'invocation-type', <awscli.arguments.CLIArgument object at 0x7f4c413bfc10>), (u'log-type', <awscli.arguments.CLIArgument object at 0x7f4c413bfc50>), (u'client-context', <awscli.arguments.CLIArgument object at 0x7f4c413bfc90>), (u'payload', <awscli.arguments.CLIArgument object at 0x7f4c413bfd10>), (u'qualifier', <awscli.arguments.CLIArgument object at 0x7f4c413bfd50>)])
2017-06-10 22:32:13,436 - MainThread - botocore.hooks - DEBUG - Event building-argument-table.lambda.invoke: calling handler <function add_streaming_output_arg at 0x7f4c417bbc08>
2017-06-10 22:32:13,436 - MainThread - botocore.hooks - DEBUG - Event building-argument-table.lambda.invoke: calling handler <function add_cli_input_json at 0x7f4c423328c0>
2017-06-10 22:32:13,436 - MainThread - botocore.hooks - DEBUG - Event building-argument-table.lambda.invoke: calling handler <function unify_paging_params at 0x7f4c41846410>
2017-06-10 22:32:13,439 - MainThread - botocore.loaders - DEBUG - Loading JSON file: /usr/local/lib/python2.7/dist-packages/botocore/data/lambda/2015-03-31/paginators-1.json
2017-06-10 22:32:13,439 - MainThread - botocore.hooks - DEBUG - Event building-argument-table.lambda.invoke: calling handler <function add_generate_skeleton at 0x7f4c418357d0>
2017-06-10 22:32:13,440 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.lambda.invoke.function-name: calling handler <function uri_param at 0x7f4c4234d578>
2017-06-10 22:32:13,440 - MainThread - botocore.hooks - DEBUG - Event process-cli-arg.lambda.invoke: calling handler <awscli.argprocess.ParamShorthandParser object at 0x7f4c4230be50>
2017-06-10 22:32:13,440 - MainThread - awscli.arguments - DEBUG - Unpacked value of u'test' for parameter "function_name": u'test'
2017-06-10 22:32:13,440 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.lambda.invoke.invocation-type: calling handler <function uri_param at 0x7f4c4234d578>
2017-06-10 22:32:13,440 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.lambda.invoke.log-type: calling handler <function uri_param at 0x7f4c4234d578>
2017-06-10 22:32:13,440 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.lambda.invoke.client-context: calling handler <function uri_param at 0x7f4c4234d578>
2017-06-10 22:32:13,440 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.lambda.invoke.payload: calling handler <function uri_param at 0x7f4c4234d578>
2017-06-10 22:32:13,440 - MainThread - botocore.hooks - DEBUG - Event process-cli-arg.lambda.invoke: calling handler <awscli.argprocess.ParamShorthandParser object at 0x7f4c4230be50>
2017-06-10 22:32:13,440 - MainThread - awscli.argprocess - DEBUG - Param payload looks like JSON, not considered for param shorthand.
2017-06-10 22:32:13,441 - MainThread - awscli.arguments - DEBUG - Unpacked value of u'{"test_sleep": "true", "test_sleep_val": 80 }' for parameter "payload": u'{"test_sleep": "true", "test_sleep_val": 80 }'
2017-06-10 22:32:13,441 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.lambda.invoke.qualifier: calling handler <function uri_param at 0x7f4c4234d578>
2017-06-10 22:32:13,441 - MainThread - botocore.hooks - DEBUG - Event load-cli-arg.lambda.invoke.outfile: calling handler <function uri_param at 0x7f4c4234d578>
2017-06-10 22:32:13,441 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: env
2017-06-10 22:32:13,441 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: assume-role
2017-06-10 22:32:13,441 - MainThread - botocore.credentials - DEBUG - Looking for credentials via: shared-credentials-file
2017-06-10 22:32:13,441 - MainThread - botocore.credentials - INFO - Found credentials in shared credentials file: ~/.aws/credentials
2017-06-10 22:32:13,442 - MainThread - botocore.loaders - DEBUG - Loading JSON file: /usr/local/lib/python2.7/dist-packages/botocore/data/endpoints.json
2017-06-10 22:32:13,460 - MainThread - botocore.client - DEBUG - Registering retry handlers for service: lambda
2017-06-10 22:32:13,461 - MainThread - botocore.hooks - DEBUG - Event creating-client-class.lambda: calling handler <function add_generate_presigned_url at 0x7f4c427b5758>
2017-06-10 22:32:13,462 - MainThread - botocore.args - DEBUG - The s3 config key is not a dictionary type, ignoring its value of: None
**2017-06-10 22:32:13,465 - MainThread - botocore.endpoint - DEBUG - Setting lambda timeout as (60, 60)**
2017-06-10 22:32:13,466 - MainThread - botocore.hooks - DEBUG - Event before-parameter-build.lambda.Invoke: calling handler <function generate_idempotent_uuid at 0x7f4c42764c80>
2017-06-10 22:32:13,467 - MainThread - botocore.endpoint - DEBUG - Making request for OperationModel(name=Invoke) (verify_ssl=True) with params: {'body': '{"test_sleep": "true", "test_sleep_val": 80 }', 'url': u'LAMBDA_URL', 'headers': {'User-Agent': 'aws-cli/1.11.93 Python/2.7.12 Linux/4.4.0-79-generic botocore/1.5.56'}, 'context': {'auth_type': None, 'client_region': 'us-east-1', 'has_streaming_input': True, 'client_config': <botocore.config.Config object at 0x7f4c41259090>}, 'query_string': {}, 'url_path': u'/2015-03-31/functions/test/invocations', 'method': u'POST'}
2017-06-10 22:32:13,468 - MainThread - botocore.hooks - DEBUG - Event request-created.lambda.Invoke: calling handler <bound method RequestSigner.handler of <botocore.signers.RequestSigner object at 0x7f4c41249fd0>>
2017-06-10 22:32:13,468 - MainThread - botocore.hooks - DEBUG - Event choose-signer.lambda.Invoke: calling handler <function set_operation_specific_signer at 0x7f4c42764b90>
2017-06-10 22:32:13,468 - MainThread - botocore.auth - DEBUG - Calculating signature using v4 auth.
2017-06-10 22:32:13,468 - MainThread - botocore.auth - DEBUG - CanonicalRequest:
POST
/2015-03-31/functions/test/invocations

host:lambda.us-east-1.amazonaws.com
x-amz-date:20170611T013213Z

host;x-amz-date
cbdb2e73182118802202540ea9c0ccd0ce3c926344f68ac555896b77c82be44c
2017-06-10 22:32:13,468 - MainThread - botocore.auth - DEBUG - StringToSign:
AWS4-HMAC-SHA256
20170611T013213Z
20170611/us-east-1/lambda/aws4_request
f19f74914bb46fa4385b875378e51686643f533d788fe2b2b08070fa241a84e7
2017-06-10 22:32:13,468 - MainThread - botocore.auth - DEBUG - Signature:
f5e103c33eba199e4c8d3768e385c3f59dea24829ac75e986c6a69dcae1e1a06
2017-06-10 22:32:13,470 - MainThread - botocore.endpoint - DEBUG - Sending http request: <PreparedRequest [POST]>
2017-06-10 22:32:13,470 - MainThread - botocore.vendored.requests.packages.urllib3.connectionpool - INFO - Starting new HTTPS connection (1): lambda.us-east-1.amazonaws.com
2017-06-10 22:33:14,030 - MainThread - botocore.endpoint - DEBUG - Exception received when sending HTTP request.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/botocore/endpoint.py", line 204, in _get_response
    proxies=self.proxies, timeout=self.timeout)
  File "/usr/local/lib/python2.7/dist-packages/botocore/vendored/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/botocore/vendored/requests/adapters.py", line 433, in send
    raise ReadTimeout(e, request=request)
ReadTimeout: HTTPSConnectionPool(host='lambda.us-east-1.amazonaws.com', port=443): Read timed out. (read timeout=60)
2017-06-10 22:33:14,058 - MainThread - botocore.hooks - DEBUG - Event needs-retry.lambda.Invoke: calling handler <botocore.retryhandler.RetryHandler object at 0x7f4c416bef90>
2017-06-10 22:33:14,058 - MainThread - botocore.retryhandler - DEBUG - retry needed, retryable exception caught: HTTPSConnectionPool(host='lambda.us-east-1.amazonaws.com', port=443): Read timed out. (read timeout=60)
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/botocore/retryhandler.py", line 269, in _should_retry
    return self._checker(attempt_number, response, caught_exception)
  File "/usr/local/lib/python2.7/dist-packages/botocore/retryhandler.py", line 317, in __call__
    caught_exception)
  File "/usr/local/lib/python2.7/dist-packages/botocore/retryhandler.py", line 223, in __call__
    attempt_number, caught_exception)
  File "/usr/local/lib/python2.7/dist-packages/botocore/retryhandler.py", line 359, in _check_caught_exception
    raise caught_exception
ReadTimeout: HTTPSConnectionPool(host='lambda.us-east-1.amazonaws.com', port=443): Read timed out. (read timeout=60)
2017-06-10 22:33:14,065 - MainThread - botocore.retryhandler - DEBUG - Retry needed, action of: 0.103661824994
2017-06-10 22:33:14,065 - MainThread - botocore.endpoint - DEBUG - Response received to retry, sleeping for 0.103661824994 seconds
2017-06-10 22:33:14,169 - MainThread - botocore.hooks - DEBUG - Event request-created.lambda.Invoke: calling handler <bound method RequestSigner.handler of <botocore.signers.RequestSigner object at 0x7f4c41249fd0>>
2017-06-10 22:33:14,170 - MainThread - botocore.hooks - DEBUG - Event choose-signer.lambda.Invoke: calling handler <function set_operation_specific_signer at 0x7f4c42764b90>
2017-06-10 22:33:14,170 - MainThread - botocore.auth - DEBUG - Calculating signature using v4 auth.
2017-06-10 22:33:14,170 - MainThread - botocore.auth - DEBUG - CanonicalRequest:
POST
/2015-03-31/functions/test/invocations

host:lambda.us-east-1.amazonaws.com
x-amz-date:20170611T013314Z

host;x-amz-date
cbdb2e73182118802202540ea9c0ccd0ce3c926344f68ac555896b77c82be44c
2017-06-10 22:33:14,171 - MainThread - botocore.auth - DEBUG - StringToSign:
AWS4-HMAC-SHA256
20170611T013314Z
20170611/us-east-1/lambda/aws4_request
6dc0d5704081a667f2cf1a562d10873a3d7c6fcdfe0f8a27f9e90bb3af4136c9
2017-06-10 22:33:14,171 - MainThread - botocore.auth - DEBUG - Signature:
2ff60d738a13a34ec2bcaf69e8f44407cafe440014a355c5977e4c9b4ce41f4f
2017-06-10 22:33:14,172 - MainThread - botocore.endpoint - DEBUG - Sending http request: <PreparedRequest [POST]>
2017-06-10 22:33:14,172 - MainThread - botocore.vendored.requests.packages.urllib3.connectionpool - INFO - Starting new HTTPS connection (2): lambda.us-east-1.amazonaws.com
^C

**Comment by jamesls** [comment]
The expected behavior is to use the default of 60 seconds for all operations.  Changing the defaults is tricky, especially for people that may be relying on the existing behavior.  To control the timeout per-invocation, you can specify the `--cli-read-timeout` option to the number of seconds to wait before timing out.

```
aws lambda invoke --cli-read-timeout 240 --function-name FUNCTION_NAME --payload 'JSON_PAYLOAD' output.json
```

Let me know if you have any more questions.

**Comment by lucasdf** [comment]
@jamesls oh I see. It took me a while to understand why the function was being invoked multiple times. 

Do you think it would make sense to have an option such as "--use-resource-timeout" that would get the resource configuration, parse the timeout and use it instead of using the default of 60 seconds?

**Comment by jamesls** [comment]
It's an interesting idea, that would require an extra API call to get that value before invoking the function.  You could write an CLI alias (https://github.com/awslabs/awscli-aliases) that did that.  It would look something like:

```
invoke-function = 
  !f() {
   aws lambda invoke ... \
     --cli-read-timeout $(aws lambda get-function-configuration --function-name $1 --query Timeout --output text)
  }; f

**Comment by lucasdf** [comment]
I wrote an alias to do that, but it seems that bash does not like arguments with quotes or double quotes, so this only works if the payload is loaded from a file and it won't work when the JSON is written inline.

`aws lambda-invoke --function-name LAMBDA_NAME --payload file://FILE_NAME output.temp`

```bash
lambda-invoke =
  !f() {
    set -e
    args="$*"

    for i in "$@"; do
      if [ "$i" = "--function-name" ]; then
        shift
        lambda_name="$1"
        break
      fi
      shift
    done
    timeout=$(aws lambda get-function-configuration --function-name $lambda_name --query Timeout --output text)
    aws lambda invoke --cli-read-timeout $timeout $args
  }; f
```

Another option, which is the one that I am using, is to use an alias to get the timeout only. 

`aws lambda invoke --function-name LAMBDA_NAME --payload PAYLOAD --cli-read-timeout $(aws lambda-timeout LAMBDA_NAME) output.temp`

```bash
lambda-timeout =
  !f() {
    aws lambda get-function-configuration --function-name "$1" --query Timeout --output text
  }; f
```

**Comment by AlJohri** [comment]
@jamesls just got burned by this- couldn't tell for the life of me while the lambda kept restarting even though it was in RequestResponse mode.

I believe this should be mentioned in the `aws lambda invoke help` page

**Comment by AlJohri** [comment]
Wow, I just got burned by this again. Tried to invoke a long running lambda from the CLI and it fired twice. Ended up sending emails to all my users twice. This is definitely not obvious from reading the command. Please change the default!

**Comment by jmehnle** [comment]
This issue was closed by a commit that "quote[s] arguments in aliases" (#2689). This merely facilitates the _work-around_ for this issue but in no way actually _fixes_ the issue. At the very least the `aws lambda invoke help` documentation should be updated to very prominently call out that if your synchronous lambda takes more than 60 seconds to return a response, it will be implicitly reinvoked every 60 seconds.

I inherited a lambda function that rotates the master user password shared by multiple RDS clusters, and this generally takes on the order of 75-120 seconds, causing an implicit reinvocation by the `aws` CLI. And it just so happened that this reinvocation interfered with the initial run, which was still in the process of updating the *first* cluster's password, which then makes both invocations fail and retry internally, which makes the `aws` CLI retry even more times. Granted, this may not be the best designed lambda (again, I inherited this), but it took me *hours* to figure out that it's the `aws` CLI that's quietly making multiple retries.

Honestly, though, this issue should be fixed regardless of documentation.

**Comment by kdaily** [comment]
@jmehnle, thanks for your comments. I'm reopening to review the behavior.

I would agree that documentation improvements would be welcome - I'll open a separate issue for that.

**Comment by hack2learn** [comment]
it would be great if the cli command could use the lambda's configured timeout.

**Comment by irvintim** [comment]
Wouldn't the easiest thing to do be to just make the default --cli-read-timeout be 900 (or maybe 910) when calling "lambda invoke"?  900 is the max lambda timeout, so it will end when the Lambda completes, or when it timeouts.

**Comment by frenchtoastbeer** [comment]
I don't want to increase the timeout, I want to disable the retry.

**Comment by jdsampayo** [comment]
@frenchtoastbeer in order to disable the retry you can use:

```
AWS_MAX_ATTEMPTS=1 aws lambda invoke ...
```
